src/apps/jsgeneric/jsgeneric.cpp

```diff
@@ -13,33 +13,101 @@ namespace ccfapp
   using namespace kv;
   using namespace ccf;
 
-  using GenericTable = ccf::Store::Map<nlohmann::json, nlohmann::json>;
+  using LogTable = ccf::Store::Map<int32_t, std::string>;
+
+  static JSValue js_print(
+    JSContext* ctx, JSValueConst this_val, int argc, JSValueConst* argv)
+  {
+    int i;
+    const char* str;
+    std::stringstream ss;
+
+    for (i = 0; i < argc; i++)
+    {
+      if (i != 0)
+        ss << ' ';
+      str = JS_ToCString(ctx, argv[i]);
+      if (!str)
+        return JS_EXCEPTION;
+      ss << str;
+      JS_FreeCString(ctx, str);
+    }
+    LOG_INFO_FMT(ss.str());
+    return JS_UNDEFINED;
+  }
+
+  void js_dump_error(JSContext* ctx)
+  {
+    JSValue exception_val = JS_GetException(ctx);
+
+    JSValue val;
+    const char* stack;
+    bool is_error;
+
+    is_error = JS_IsError(ctx, exception_val);
+    if (!is_error)
+      LOG_INFO_FMT("Throw: ");
+    js_print(ctx, JS_NULL, 1, (JSValueConst*)&exception_val);
+    if (is_error)
+    {
+      val = JS_GetPropertyStr(ctx, exception_val, "stack");
+      if (!JS_IsUndefined(val))
+      {
+        stack = JS_ToCString(ctx, val);
+        LOG_INFO_FMT("{}", stack);
+
+        JS_FreeCString(ctx, stack);
+      }
+      JS_FreeValue(ctx, val);
+    }
+
+    JS_FreeValue(ctx, exception_val);
+  }
+
+  static JSValue js_get(
+    JSContext* ctx, JSValueConst this_val, int argc, JSValueConst* argv)
+  {
+    auto log_table_view = (LogTable::TxView*)JS_GetContextOpaque(ctx);
+    if (!JS_IsInteger(argv[0]))
+      return JS_EXCEPTION;
+    int32_t i = JS_VALUE_GET_INT(argv[0]);
+    auto str = log_table_view->get(i);
+    if (str.has_value())
+      return JS_NewStringLen(ctx, str.value().data(), str.value().size());
+    else
+      return JS_EXCEPTION;
+  }
+
+  static JSValue js_put(
+    JSContext* ctx, JSValueConst this_val, int argc, JSValueConst* argv)
+  {
+    auto log_table_view = (LogTable::TxView*)JS_GetContextOpaque(ctx);
+    if (!JS_IsInteger(argv[0]))
+      return JS_EXCEPTION;
+    int32_t i = JS_VALUE_GET_INT(argv[0]);
+    auto v = JS_ToCString(ctx, argv[1]); // TODO: error checking?
+    std::string s(v);
```

Comment:
Could we use a `string_view` here to avoid the copy?

Reply:
I think we can skip the intermediary altogether, yes.